Document Element

jknaudt21 opened the pull request:
### Description of Change

Manually add the rest of `Element` and migrated old .xml documents belonging to this class into the source.

There were some _peculiar_ aspects about this class, however. Empty functions that should (?) invoke events but instead leave it up the derived classes to do so. These are:
- OnParentChanging
- OnParentChanged
- OnHandlerChanging
- OnHandlerChanged

jknaudt21 committed:
Begin doc progress

jknaudt21 committed:
Document element

jknaudt21 committed:
More explicit interface implementations

jfversluis commented on src/Controls/docs/Microsoft.Maui.Controls/Element.xml:
```diff
@@ -313,93 +162,13 @@
         <ReturnType>System.Collections.Generic.IEnumerable&lt;Microsoft.Maui.Controls.Element&gt;</ReturnType>
       </ReturnValue>
-      <Parameters />
+      <Parameters />EffectIsAttached
```
This seems like some accidental paste 😄

jfversluis commented on src/Controls/docs/Microsoft.Maui.Controls/Element.xml:
```diff
@@ -1014,9 +408,9 @@
-        <summary>Gets or sets a user defined value to uniquely identify the element.</summary>
-        <value>A string uniquely identifying the element.</value>
-        <remarks>Use the StyleId property to identify individual elements in your application for identification in ui testing and in theme engines.</remarks>
+<summary>Gets or sets a user defined value to uniquely identify the element.</summary>
+<value>A string uniquely identifying the element.</value>
+<remarks>Use the StyleId property to identify individual elements in your application for identification in ui testing and in theme engines.</remarks>
```
Was this supposed to happen?

jknaudt21 replied:
Meant to delete entry entry. Good catch

jfversluis commented on src/Controls/src/Core/Element/Element.cs:
```diff
@@ -43,7 +43,9 @@ public abstract partial class Element : BindableObject, IElementDefinition, INam
 
-		/// <include file="../../docs/Microsoft.Maui.Controls/Element.xml" path="//Member[@MemberName='AutomationId']/Docs/*" />
+		/// <summary>Gets or sets a value that allows the automation framework to find and interact with this element.</summary>
+		/// <value>A value that the automation framework can use to find and interact with this element.</value>
```
Personally I don’t think I used the Value tag too much. Is there any specific reason you do? Does it give us any extras?

jknaudt21 replied:
Personally, I didn't find super useful. _**But**_, it was being used in the old xml docs, so I figured I might as well copy it over, and the API docs guide implies you should use it: https://github.com/dotnet/dotnet-api-docs/wiki/Property-Value.

It also reflects on MSLearn  adding a section "Property Value". Happy to remove the tag if you think that's the best way to go about it

davidbritch replied:
My view: I think it's important to have the <value> entry. It's just that I don't think the current text is that useful - it essentially repeats the text from the summary.

jfversluis commented on src/Controls/src/Core/Element/Element.cs:
```diff
@@ -56,14 +58,17 @@ public string AutomationId
 		}
 
-		/// <include file="../../docs/Microsoft.Maui.Controls/Element.xml" path="//Member[@MemberName='Effects']/Docs/*" />
+		/// <summary> Gets or sets the styles and properties that will be applied to the element during runtime.</summary>
+		/// <value> A collection containing the different <see cref="Effect"/> to be applied to the element</value>
```
```suggestion
		/// <summary>Gets or sets the styles and properties that will be applied to the element during runtime.</summary>
		/// <value>A collection containing the different <see cref="Effect"/> to be applied to the element</value>
```

davidbritch replied:
Add a period to the end of the sentence for the value description.

jfversluis commented on src/Controls/src/Core/Element/Element.cs:
```diff
@@ -89,7 +97,9 @@ public Guid Id
-		/// <include file="../../docs/Microsoft.Maui.Controls/Element.xml" path="//Member[@MemberName='StyleId']/Docs/*" />
+		/// <summary>Gets or sets a user defined value to uniquely identify the element.</summary>
+		/// <value>A string uniquely identifying the element.</value>
+		/// <remarks>Use the StyleId property to identify individual elements in your application for identification in ui testing and in theme engines.</remarks>
```
```suggestion
		/// <remarks>Use the StyleId property to identify individual elements in your application for identification in UI testing and in theme engines.</remarks>
```

UI? Also can we add a <see cref=“StyleId”/> for the reference? (On my phone so can’t check all things sorry 😬)

I think this applies to some more places; both the UI spelling and some more references to types

davidbritch replied:
It should definitely be UI rather than ui (in all comments).

davidbritch replied:
Add a cref to StyleId in the remarks?

jfversluis commented on src/Controls/src/Core/Element/Element.cs:
```diff
@@ -266,20 +277,22 @@ internal Element ParentOverride
 			}
 		}
 
-		// you're not my real dad
```
😅

jfversluis commented on src/Controls/src/Core/Element/Element.cs:
```diff
@@ -266,20 +277,22 @@ internal Element ParentOverride
 
-		// you're not my real dad
-		/// <include file="../../docs/Microsoft.Maui.Controls/Element.xml" path="//Member[@MemberName='RealParent']/Docs/*" />
+		/// <summary>For internal use by the Microsoft.Maui.Controls platform.</summary>
```
Can we describe what this actually does and add this as a remark?

Its a public API, people should be able to use it 🤷‍♂️ 

Same for a couple of others below

jknaudt21 replied:
I spoke about this with @PureWeen. They should be made internal, but that will cause compile issues. I wanted to keep the scope of the PR to only docs.

jknaudt21 replied:
Maybe we can mark all these internal things as obsolete too?

davidbritch replied:
`[EditorBrowsable(EditorBrowsableState.Never)]` means the API won't show up in intellisense. However, it will show up in the API docs. Not sure it's worth doc'ing something that only appears in one of the two information sources.

jfversluis reviewed (commented):
Some quick comments from Italy, ciao!

davidbritch commented on src/Controls/src/Core/Element/Element.cs:
```diff
@@ -139,6 +149,7 @@ private protected virtual IList<Element> LogicalChildrenInternalBackingStore
 		public ReadOnlyCollection<Element> LogicalChildren =>
 			new ReadOnlyCollection<Element>(new TemporaryWrapper(LogicalChildrenInternal));
 
+		/// <inheritdoc/>
```
It'll be interesting to see if mdoc is smart enough to pick this up, when it's not on the public declaration.

davidbritch commented on src/Controls/src/Core/Element/Element.cs:
```diff
@@ -78,7 +83,10 @@ public IList<Effect> Effects
 
-		/// <include file="../../docs/Microsoft.Maui.Controls/Element.xml" path="//Member[@MemberName='Id']/Docs/*" />
+		/// <summary>Gets a value that can be used to uniquely identify an element throughout the run of your application.</summary>
+		/// <value>A Guid uniquely identifying the element.</value>
```
Shouldn't this be a type link to System.Guid?

davidbritch commented on src/Controls/src/Core/Element/Element.cs:
```diff
@@ -391,7 +407,10 @@ public bool EffectIsAttached(string name)
 		}
 
-		/// <include file="../../docs/Microsoft.Maui.Controls/Element.xml" path="//Member[@MemberName='FindByName']/Docs/*" />
+		/// <summary> Returns the element that has the specified name.</summary>
```
```suggestion
		/// <summary>Returns the element that has the specified name.</summary>
```

davidbritch commented on src/Controls/src/Core/Element/Element.cs:
```diff
@@ -391,7 +407,10 @@ public bool EffectIsAttached(string name)
 
-		/// <include file="../../docs/Microsoft.Maui.Controls/Element.xml" path="//Member[@MemberName='FindByName']/Docs/*" />
+		/// <summary> Returns the element that has the specified name.</summary>
+		/// <param name="name">The name of the element to be found</param>
```
```suggestion
		/// <param name="name">The name of the element to be found.</param>
```

davidbritch commented on src/Controls/src/Core/Element/Element.cs:
```diff
@@ -400,43 +419,56 @@ public object FindByName(string name)
 			namescope.UnregisterName(name);
 		}
 
+		/// <summary>Occurs whenever a child element is added to the element.</summary>
```
```suggestion
		/// <summary>Raised whenever a child element is added to the element.</summary>
```

davidbritch commented on src/Controls/src/Core/Element/Element.cs:
```diff
@@ -400,43 +419,56 @@ public object FindByName(string name)
+		/// <summary>Occurs whenever a child element is added to the element.</summary>
 		public event EventHandler<ElementEventArgs> ChildAdded;
 
+		/// <summary>Occurs whenever a child element is removed from the element.</summary>
```
```suggestion
		/// <summary>Raised whenever a child element is removed from the element.</summary>
```

davidbritch commented on src/Controls/src/Core/Element/Element.cs:
```diff
@@ -400,43 +419,56 @@ public object FindByName(string name)
+		/// <summary>Occurs whenever a child element is removed from the element.</summary>
 		public event EventHandler<ElementEventArgs> ChildRemoved;
 
+		/// <summary>Occurs whenever a child element is added to the element's subtree.</summary>
```
```suggestion
		/// <summary>Raised whenever a child element is added to the element's subtree.</summary>
```

davidbritch commented on src/Controls/src/Core/Element/Element.cs:
```diff
@@ -400,43 +419,56 @@ public object FindByName(string name)
+		/// <summary>Occurs whenever a child element is added to the element's subtree.</summary>
 		public event EventHandler<ElementEventArgs> DescendantAdded;
 
+		/// <summary>Occurs whenever a child element is removed from the elements subtree.</summary>
```
```suggestion
		/// <summary>Raised whenever a child element is removed from the elements subtree.</summary>
```

davidbritch commented on src/Controls/src/Core/Element/Element.cs:
```diff
@@ -400,43 +419,56 @@ public object FindByName(string name)
 
-		/// <include file="../../docs/Microsoft.Maui.Controls/Element.xml" path="//Member[@MemberName='RemoveDynamicResource']/Docs/*" />
+		/// <summary>Removes a previously set dynamic resource.</summary>
+		/// <param name="property">The BindableProperty from which to remove the DynamicResource</param>
```
```suggestion
		/// <param name="property">The BindableProperty from which to remove the DynamicResource.</param>
```

davidbritch replied:
BindableProperty could be a cref.

davidbritch commented on src/Controls/src/Core/Element/Element.cs:
```diff
@@ -400,43 +419,56 @@ public object FindByName(string name)
 
-		/// <include file="../../docs/Microsoft.Maui.Controls/Element.xml" path="//Member[@MemberName='SetDynamicResource']/Docs/*" />
+		/// <summary>Sets the <see cref="BindableProperty"/> property of this element to be updated via the DynamicResource with the provided key.</summary>
+		/// <param name="property">The property to be updated</param>
```
```suggestion
		/// <param name="property">The property to be updated.</param>
```

davidbritch replied:
```suggestion
		/// <param name="property">The property to be updated.</param>
```

davidbritch commented on src/Controls/src/Core/Element/Element.cs:
```diff
@@ -400,43 +419,56 @@ public object FindByName(string name)
-		/// <include file="../../docs/Microsoft.Maui.Controls/Element.xml" path="//Member[@MemberName='SetDynamicResource']/Docs/*" />
+		/// <summary>Sets the <see cref="BindableProperty"/> property of this element to be updated via the DynamicResource with the provided key.</summary>
+		/// <param name="property">The property to be updated</param>
+		/// <param name="key">The key for the requested resource</param>
```
```suggestion
		/// <param name="key">The key for the requested resource.</param>
```

davidbritch commented on src/Controls/src/Core/Element/Element.cs:
```diff
@@ -453,6 +485,9 @@ protected override void OnBindingContextChanged()
 
+		/// <summary>Raises the <see cref="ChildAdded"/> event. Implement this method to add class handling for this event.</summary>
+		/// <remarks>This method has no default implementation. You should still call the base implementation in case an intermediate class has implemented this method.</remarks>
+		/// <param name="child">The element that's been added as a child</param>
```
```suggestion
		/// <param name="child">The element that's been added as a child.</param>
```

davidbritch commented on src/Controls/src/Core/Element/Element.cs:
```diff
@@ -468,6 +503,13 @@ protected virtual void OnChildAdded(Element child)
+		/// <summary> Raises the <see cref="ChildRemoved"/> event. Implement this method to add class handling for this event </summary>
+		/// <remarks>
+		/// This method has no default implementation. You should still call the base implementation in case an intermediate class has implemented this method.
+		/// If not debugging, the logical tree index will not have any effect
```
```suggestion
		/// If not debugging, the logical tree index will not have any effect.
```

davidbritch commented on src/Controls/src/Core/Element/Element.cs:
```diff
@@ -468,6 +503,13 @@ protected virtual void OnChildAdded(Element child)
+		/// This method has no default implementation. You should still call the base implementation in case an intermediate class has implemented this method.
+		/// If not debugging, the logical tree index will not have any effect
+		/// </remarks>
+		/// <param name="child">The child element that's been removed</param>
```
```suggestion
		/// <param name="child">The child element that's been removed.</param>
```

davidbritch commented on src/Controls/src/Core/Element/Element.cs:
```diff
@@ -468,6 +503,13 @@ protected virtual void OnChildAdded(Element child)
+		/// If not debugging, the logical tree index will not have any effect
+		/// </remarks>
+		/// <param name="child">The child element that's been removed</param>
+		/// <param name="oldLogicalIndex">The child's element index in the logical tree</param>
```
```suggestion
		/// <param name="oldLogicalIndex">The child's element index in the logical tree.</param>
```

davidbritch commented on src/Controls/src/Core/Element/Element.cs:
```diff
@@ -480,14 +522,17 @@ protected virtual void OnChildRemoved(Element child, int oldLogicalIndex)
 				OnDescendantRemoved(element);
 		}
-
+		/// <summary>Raises the <see cref="ParentSet"/> event. Implement this method in order to add behavior when the element is added to a parent</summary>
```
```suggestion
		/// <summary>Raises the <see cref="ParentSet"/> event. Implement this method in order to add behavior when the element is added to a parent.</summary>
```

davidbritch commented on src/Controls/src/Core/Element/Element.cs:
```diff
@@ -761,20 +815,34 @@ private protected virtual void OnParentChangingCore(Element oldParent, Element n
 			set => SetHandler(value);
 		}
 
+		/// <summary>Occurs whenever the element's handler starts to change.</summary>
```
```suggestion
		/// <summary>Raised whenever the element's handler starts to change.</summary>
```

davidbritch commented on src/Controls/src/Core/Element/Element.cs:
```diff
@@ -761,20 +815,34 @@ private protected virtual void OnParentChangingCore(Element oldParent, Element n
+		/// <param name="args">Provides data for the <see cref="HandlerChanging"/> event</param>
 		protected virtual void OnHandlerChanging(HandlerChangingEventArgs args) { }
 
+		/// <summary>When overridden in a derived class, should raise the <see cref="HandlerChanged"/> event</summary>
```
```suggestion
		/// <summary>When overridden in a derived class, should raise the <see cref="HandlerChanged"/> event.</summary>
```

davidbritch commented on src/Controls/src/Core/Element/Element.cs:
```diff
@@ -761,20 +815,34 @@ private protected virtual void OnParentChangingCore(Element oldParent, Element n
 		protected virtual void OnHandlerChanging(HandlerChangingEventArgs args) { }
 
+		/// <summary>When overridden in a derived class, should raise the <see cref="HandlerChanged"/> event</summary>
+		/// <remarks>It is the implementor's responsibility to raise the <see cref="HandlerChanged"/> event</remarks>
```
```suggestion
		/// <remarks>It is the implementor's responsibility to raise the <see cref="HandlerChanged"/> event.</remarks>
```

davidbritch commented on src/Controls/src/Core/Element/Element.cs:
```diff
@@ -761,20 +815,34 @@ private protected virtual void OnParentChangingCore(Element oldParent, Element n
 
+		/// <summary>When overridden in a derived class, should raise the <see cref="HandlerChanging"/> event</summary>
+		/// <remarks>It is the implementor's responsibility to raise the <see cref="HandlerChanging"/> event</remarks>
+		/// <param name="args">Provides data for the <see cref="HandlerChanging"/> event</param>
```
```suggestion
		/// <param name="args">Provides data for the <see cref="HandlerChanging"/> event.</param>
```

davidbritch commented on src/Controls/src/Core/Element/Element.cs:
```diff
@@ -761,20 +815,34 @@ private protected virtual void OnParentChangingCore(Element oldParent, Element n
 		public event EventHandler HandlerChanged;
 
+		/// <summary>When overridden in a derived class, should raise the <see cref="HandlerChanging"/> event</summary>
+		/// <remarks>It is the implementor's responsibility to raise the <see cref="HandlerChanging"/> event</remarks>
```
```suggestion
		/// <remarks>It is the implementor's responsibility to raise the <see cref="HandlerChanging"/> event.</remarks>
```

davidbritch commented on src/Controls/src/Core/Element/Element.cs:
```diff
@@ -761,20 +815,34 @@ private protected virtual void OnParentChangingCore(Element oldParent, Element n
+		/// <summary>Occurs whenever the element's handler has changed.</summary>
 		public event EventHandler HandlerChanged;
 
+		/// <summary>When overridden in a derived class, should raise the <see cref="HandlerChanging"/> event</summary>
```
```suggestion
		/// <summary>When overridden in a derived class, should raise the <see cref="HandlerChanging"/> event.</summary>
```

davidbritch reviewed (changes requested):
You can delete Element.xml if you add inline /// at the class level.

davidbritch commented on src/Controls/src/Core/Element/Element.cs:
```diff
@@ -761,20 +815,34 @@ private protected virtual void OnParentChangingCore(Element oldParent, Element n
+		/// <summary>Gets or sets the associated handler for this element.</summary>
+		/// <value>An implementation of <see cref="IElementHandler"/></value>
+		/// <remarks>Maps the element to platform-specific controls and implementations.</remarks>
+		/// <seealso href="https://learn.microsoft.com/en-us/dotnet/maui/user-interface/handlers/">Conceptual documentation on handlers</seealso>
```
Only just considered intellisense, which will require the fully qualified URL. Regardless, the URL shouldn't specify the locale.

jknaudt21 replied:
Good point. Should I make a fwdlink that directs just to https://learn.microsoft.com/dotnet/maui/user-interface/handlers/?

davidbritch replied:
It's not worth it. Just using the full URL (without locale) is fine. If I ever move the content I'll put a redirect in place.

jknaudt21 committed:
Address PR feedback

jknaudt21 committed:
Document interfaces + delete xml

jknaudt21 committed:
whitespace

jknaudt21 commented on src/Controls/src/Core/Element/Element.cs:
```diff
@@ -761,20 +848,34 @@ private protected virtual void OnParentChangingCore(Element oldParent, Element n
+		/// <summary>Gets or sets the associated handler for this element.</summary>
+		/// <value>An implementation of <see cref="IElementHandler"/>.</value>
+		/// <remarks>Maps the element to platform-specific controls and implementations.</remarks>
+		/// <seealso href="https://learn.microsoft.com/dotnet/maui/user-interface/handlers/">Conceptual documentation on handlers</seealso>
```
@davidbritch is this link good?

davidbritch replied:
Perfect!

jknaudt21 committed:
Prevent compile issues

jknaudt21 committed:
Document IElementController (and mark as obsolete)

davidbritch reviewed (approved):
LGTM!

jknaudt21 committed:
Remove Obsolete attribute from IElementController

This would break the build in various places - and would go out of scope of the PR

jknaudt21 committed:
Document IVusalElementController

Apparently is only for internal use. We can't make it private so the only choice is to mark everything as for internal use only

davidbritch commented on src/Controls/src/Core/IVisualElementController.cs:
```diff
@@ -5,18 +5,40 @@
+	/// <summary>For internal use by the <see cref="Controls"/> platform.</summary>
 	public interface IVisualElementController : IElementController
 	{
+		/// <summary>For internal use by the <see cref="Controls"/> platform.</summary>
```
I'm not convinced that mdoc will translate Controls into Microsoft.Maui.Controls. Happy to be proven wrong though.

jknaudt21 replied:
I updated that cref to be  Microsoft.Maui.Controls

jknaudt21 committed:
Prevent mdoc from breaking

You can't really cref a namespace

jknaudt21 committed:
Fixed other crefs

jknaudt21 committed:
Address Dave's feedback

Eilon reviewed (approved):
This looks like an improvement on the docs, so can we merge it as-is? Any reason not to merge it now?

Eilon commented:
Windows build is failing with:

> D:\a\_work\1\s\bin\dotnet\sdk\7.0.203\Microsoft.Common.CurrentVersion.targets(4685,5): error MSB3027: Could not copy "obj\Release\netstandard2.0\Microsoft.Maui.Graphics.Text.Markdig.dll" to "bin\Release\netstandard2.0\Microsoft.Maui.Graphics.Text.Markdig.dll". Exceeded retry count of 10. Failed.  [D:\a\_work\1\s\src\Graphics\src\Text.Markdig\Graphics.Text.Markdig.csproj]
D:\a\_work\1\s\bin\dotnet\sdk\7.0.203\Microsoft.Common.CurrentVersion.targets(4685,5): error MSB3021: Unable to copy file "obj\Release\netstandard2.0\Microsoft.Maui.Graphics.Text.Markdig.dll" to "bin\Release\netstandard2.0\Microsoft.Maui.Graphics.Text.Markdig.dll". The requested operation cannot be performed on a file with a user-mapped section open. : 'D:\a\_work\1\s\src\Graphics\src\Text.Markdig\bin\Release\netstandard2.0\Microsoft.Maui.Graphics.Text.Markdig.dll' [D:\a\_work\1\s\src\Graphics\src\Text.Markdig\Graphics.Text.Markdig.csproj]

It's re-running for attempt 6 now... this does seem like a random failure, and probably worth its own investigation, but probably not something to block this PR on.

jfversluis commented:
Rebased to fix the build errors, all looks good now! Ship it!

jfversluis commented:
Suggesting backport to also get the benefits in .NET 7 from this. Should be relatively easy.
